DynamoDB execute_statement - TypeError: Object of type Decimal is not JSON serializable

## Overview

While testing code that mocks DynamoDB's `execute_statement`, I noticed it's having issues with items containing number types. I get the following error: `TypeError: Object of type Decimal is not JSON serializable`. I know boto3 deserializes numbers as Decimals and I see moto is trying to serialize the dictionary, but it fails since Decimals aren't JSON serializable. The error seems to always occur no matter what kind of number is added to the table.

Given moto's `execute_statement` is [experimental](https://docs.getmoto.org/en/latest/docs/services/dynamodb.html), this might be a bug with how the PartiQL parser is currently implemented.

## Environment

I'm running Python 3.8.10 on Ubuntu in Windows Subsystem for Linux (WSL) 2. I used Python mocks when using moto 4.2.5. Below is my `requirements.txt` file:

```
boto3==1.28.62
botocore==1.31.62
certifi==2023.7.22
cffi==1.16.0
charset-normalizer==3.3.0
cryptography==41.0.4
docker==6.1.3
exceptiongroup==1.1.3
idna==3.4
iniconfig==2.0.0
Jinja2==3.1.2
jmespath==1.0.1
MarkupSafe==2.1.3
moto==4.2.5
packaging==23.2
pluggy==1.3.0
py-partiql-parser==0.3.7
pycparser==2.21
pytest==7.4.2
python-dateutil==2.8.2
PyYAML==6.0.1
requests==2.31.0
responses==0.23.3
s3transfer==0.7.0
six==1.16.0
tomli==2.0.1
types-PyYAML==6.0.12.12
urllib3==1.26.16
websocket-client==1.6.4
werkzeug==3.0.0
xmltodict==0.13.0
```

## Steps to Reproduce

I created a virtual environment and installed all the dependencies above. Then I created the following files and ran `pytest` to run the test:

**conftest.py**
```python
import boto3
from moto import mock_dynamodb
import os
import pytest


@pytest.fixture
def aws_credentials():
    """Mocked AWS Credentials for moto."""
    os.environ["AWS_ACCESS_KEY_ID"] = "testing"
    os.environ["AWS_SECRET_ACCESS_KEY"] = "testing"
    os.environ["AWS_SECURITY_TOKEN"] = "testing"
    os.environ["AWS_SESSION_TOKEN"] = "testing"
    os.environ["AWS_DEFAULT_REGION"] = "us-east-1"


@pytest.fixture
def dynamodb_client(aws_credentials):
    with mock_dynamodb():
        yield boto3.client("dynamodb")


@pytest.fixture
def table_name():
    return "test-table"


@pytest.fixture
def dynamodb_table(dynamodb_client, table_name):
    # Create a mock DynamoDB table with items
    dynamodb_client.create_table(
        TableName=table_name,
        KeySchema=[{"AttributeName": "Id", "KeyType": "HASH"}],
        AttributeDefinitions=[{"AttributeName": "Id", "AttributeType": "S"}],
        BillingMode="PAY_PER_REQUEST",  # don't specify RCUs or WCUs
    )

    items = [
        {
            "Id": {
                "S": "0",
            },
            "Name": {
                "S": "Lambda",
            },
            "NameLower": {"S": "lambda"},
            "Description": {
                "S": "Run code in under 15 minutes",
            },
            "DescriptionLower": {"S": "run code in under 15 minutes"},
            "Price": {
                "N": "2E-7",
            },
            "Unit": {
                "S": "invocation",
            },
            "Category": {
                "S": "free",
            },
            "FreeTier": {
                "N": "1E+6",
            },
        },
        {
            "Id": {
                "S": "1",
            },
            "Name": {
                "S": "Auto Scaling",
            },
            "NameLower": {"S": "auto scaling"},
            "Description": {
                "S": "Automatically scale the number of EC2 instances with demand",
            },
            "DescriptionLower": {
                "S": "automatically scale the number of ec2 instances with demand"
            },
            "Price": {
                "N": "0",
            },
            "Unit": {
                "S": "group",
            },
            "Category": {
                "S": "free",
            },
            "FreeTier": {
                "NULL": True,
            },
        },
        {
            "Id": {
                "S": "2",
            },
            "Name": {
                "S": "EC2",
            },
            "NameLower": {"S": "ec2"},
            "Description": {
                "S": "Servers in the cloud",
            },
            "DescriptionLower": {"S": "servers in the cloud"},
            "Price": {
                "N": "7.2",
            },
            "Unit": {
                "S": "instance",
            },
            "Category": {
                "S": "trial",
            },
        },
        {
            "Id": {
                "S": "3",
            },
            "Name": {
                "S": "Config",
            },
            "NameLower": {"S": "config"},
            "Description": {
                "S": "Audit the configuration of AWS resources",
            },
            "DescriptionLower": {
                "S": "audit the configuration of aws resources",
            },
            "Price": {
                "N": "0.003",
            },
            "Unit": {
                "S": "configuration item",
            },
            "Category": {
                "S": "paid",
            },
        },
    ]

    for item in items:
        dynamodb_client.put_item(TableName=table_name, Item=item)

    return items
```

**test_helpers.py**
```python
def test_query_services_with_all_params(dynamodb_client, dynamodb_table, table_name):
    partiql_statement = f"SELECT * FROM \"{table_name}\" WHERE (contains(\"NameLower\", 'code') OR contains(\"DescriptionLower\", 'code')) AND Category = 'free' AND Price >= 0 AND Price <= 1 AND FreeTier IS NOT MISSING AND attribute_type(\"FreeTier\", 'N')"
    response = dynamodb_client.execute_statement(Statement=partiql_statement)
    items = response["Items"]

    assert len(items) == 1
    assert items[0] == next(
        item for item in dynamodb_table if item["Name"]["S"] == "Lambda"
    )
```

## Expected Result

With the above PartiQL query, I expect `execute_statement` to return the "Lambda" item in the DynamoDB table since it satisfies all the conditions. In other words, `items` in `test_helpers.py` should return:

```python
 [
        {
            "Id": {
                "S": "0",
            },
            "Name": {
                "S": "Lambda",
            },
            "NameLower": {"S": "lambda"},
            "Description": {
                "S": "Run code in under 15 minutes",
            },
            "DescriptionLower": {"S": "run code in under 15 minutes"},
            "Price": {
                "N": "2E-7",
            },
            "Unit": {
                "S": "invocation",
            },
            "Category": {
                "S": "free",
            },
            "FreeTier": {
                "N": "1E+6",
            },
        },
]
```

## Actual Result

When I run `pytest`, I get the following error instead:

```
====================================================== test session starts ======================================================
platform linux -- Python 3.8.10, pytest-7.4.2, pluggy-1.3.0
rootdir: /home/abhiek187/repro
collected 1 item                                                                                                                

test_helpers.py F                                                                                                         [100%]

=========================================================== FAILURES ============================================================
______________________________________________ test_query_services_with_all_params ______________________________________________

dynamodb_client = <botocore.client.DynamoDB object at 0x7f3dee3bfac0>
dynamodb_table = [{'Category': {'S': 'free'}, 'Description': {'S': 'Run code in under 15 minutes'}, 'DescriptionLower': {'S': 'run code...tion of AWS resources'}, 'DescriptionLower': {'S': 'audit the configuration of aws resources'}, 'Id': {'S': '3'}, ...}]
table_name = 'test-table'

    def test_query_services_with_all_params(dynamodb_client, dynamodb_table, table_name):
        partiql_statement = f"SELECT * FROM \"{table_name}\" WHERE (contains(\"NameLower\", 'code') OR contains(\"DescriptionLower\", 'code')) AND Category = 'free' AND Price >= 0 AND Price <= 1 AND FreeTier IS NOT MISSING AND attribute_type(\"FreeTier\", 'N')"
>       response = dynamodb_client.execute_statement(Statement=partiql_statement)

test_helpers.py:3: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
venv/lib/python3.8/site-packages/botocore/client.py:535: in _api_call
    return self._make_api_call(operation_name, kwargs)
venv/lib/python3.8/site-packages/botocore/client.py:963: in _make_api_call
    http, parsed_response = self._make_request(
venv/lib/python3.8/site-packages/botocore/client.py:986: in _make_request
    return self._endpoint.make_request(operation_model, request_dict)
venv/lib/python3.8/site-packages/botocore/endpoint.py:119: in make_request
    return self._send_request(request_dict, operation_model)
venv/lib/python3.8/site-packages/botocore/endpoint.py:202: in _send_request
    while self._needs_retry(
venv/lib/python3.8/site-packages/botocore/endpoint.py:354: in _needs_retry
    responses = self._event_emitter.emit(
venv/lib/python3.8/site-packages/botocore/hooks.py:412: in emit
    return self._emitter.emit(aliased_event_name, **kwargs)
venv/lib/python3.8/site-packages/botocore/hooks.py:256: in emit
    return self._emit(event_name, kwargs)
venv/lib/python3.8/site-packages/botocore/hooks.py:239: in _emit
    response = handler(**kwargs)
venv/lib/python3.8/site-packages/botocore/retryhandler.py:207: in __call__
    if self._checker(**checker_kwargs):
venv/lib/python3.8/site-packages/botocore/retryhandler.py:284: in __call__
    should_retry = self._should_retry(
venv/lib/python3.8/site-packages/botocore/retryhandler.py:307: in _should_retry
    return self._checker(
venv/lib/python3.8/site-packages/botocore/retryhandler.py:363: in __call__
    checker_response = checker(
venv/lib/python3.8/site-packages/botocore/retryhandler.py:247: in __call__
    return self._check_caught_exception(
venv/lib/python3.8/site-packages/botocore/retryhandler.py:416: in _check_caught_exception
    raise caught_exception
venv/lib/python3.8/site-packages/botocore/endpoint.py:278: in _do_get_response
    responses = self._event_emitter.emit(event_name, request=request)
venv/lib/python3.8/site-packages/botocore/hooks.py:412: in emit
    return self._emitter.emit(aliased_event_name, **kwargs)
venv/lib/python3.8/site-packages/botocore/hooks.py:256: in emit
    return self._emit(event_name, kwargs)
venv/lib/python3.8/site-packages/botocore/hooks.py:239: in _emit
    response = handler(**kwargs)
venv/lib/python3.8/site-packages/moto/core/botocore_stubber.py:61: in __call__
    status, headers, body = response_callback(
venv/lib/python3.8/site-packages/moto/core/responses.py:245: in dispatch
    return cls()._dispatch(*args, **kwargs)
venv/lib/python3.8/site-packages/moto/core/responses.py:441: in _dispatch
    return self.call_action()
venv/lib/python3.8/site-packages/moto/utilities/aws_headers.py:46: in _wrapper
    response = f(*args, **kwargs)
venv/lib/python3.8/site-packages/moto/utilities/aws_headers.py:74: in _wrapper
    response = f(*args, **kwargs)
venv/lib/python3.8/site-packages/moto/dynamodb/responses.py:197: in call_action
    response = getattr(self, endpoint)()
venv/lib/python3.8/site-packages/moto/dynamodb/responses.py:1108: in execute_statement
    items = self.dynamodb_backend.execute_statement(
venv/lib/python3.8/site-packages/moto/dynamodb/models/__init__.py:791: in execute_statement
    [json.dumps(item.to_regular_json()) for item in table.all_items()]
venv/lib/python3.8/site-packages/moto/dynamodb/models/__init__.py:791: in <listcomp>
    [json.dumps(item.to_regular_json()) for item in table.all_items()]
/usr/lib/python3.8/json/__init__.py:231: in dumps
    return _default_encoder.encode(obj)
/usr/lib/python3.8/json/encoder.py:199: in encode
    chunks = self.iterencode(o, _one_shot=True)
/usr/lib/python3.8/json/encoder.py:257: in iterencode
    return _iterencode(o, 0)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <json.encoder.JSONEncoder object at 0x7f3df0f70790>, o = Decimal('2E-7')

    def default(self, o):
        """Implement this method in a subclass such that it returns
        a serializable object for ``o``, or calls the base implementation
        (to raise a ``TypeError``).
    
        For example, to support arbitrary iterators, you could
        implement default like this::
    
            def default(self, o):
                try:
                    iterable = iter(o)
                except TypeError:
                    pass
                else:
                    return list(iterable)
                # Let the base class default method raise the TypeError
                return JSONEncoder.default(self, o)
    
        """
>       raise TypeError(f'Object of type {o.__class__.__name__} '
                        f'is not JSON serializable')
E       TypeError: Object of type Decimal is not JSON serializable

/usr/lib/python3.8/json/encoder.py:179: TypeError
==================================================== short test summary info ====================================================
FAILED test_helpers.py::test_query_services_with_all_params - TypeError: Object of type Decimal is not JSON serializable
======================================================= 1 failed in 0.76s =======================================================
```

--- a/setup.cfg
+++ b/setup.cfg
@@ -52,7 +52,7 @@ all =
     openapi-spec-validator>=0.2.8
     pyparsing>=3.0.7
     jsondiff>=1.1.2
-    py-partiql-parser==0.4.0
+    py-partiql-parser==0.4.1
     aws-xray-sdk!=0.96,>=0.93
     setuptools
     multipart
@@ -67,7 +67,7 @@ proxy =
     openapi-spec-validator>=0.2.8
     pyparsing>=3.0.7
     jsondiff>=1.1.2
-    py-partiql-parser==0.4.0
+    py-partiql-parser==0.4.1
     aws-xray-sdk!=0.96,>=0.93
     setuptools
     multipart
@@ -82,7 +82,7 @@ server =
     openapi-spec-validator>=0.2.8
     pyparsing>=3.0.7
     jsondiff>=1.1.2
-    py-partiql-parser==0.4.0
+    py-partiql-parser==0.4.1
     aws-xray-sdk!=0.96,>=0.93
     setuptools
     flask!=2.2.0,!=2.2.1
@@ -116,7 +116,7 @@ cloudformation =
     openapi-spec-validator>=0.2.8
     pyparsing>=3.0.7
     jsondiff>=1.1.2
-    py-partiql-parser==0.4.0
+    py-partiql-parser==0.4.1
     aws-xray-sdk!=0.96,>=0.93
     setuptools
 cloudfront =
@@ -139,10 +139,10 @@ dms =
 ds = sshpubkeys>=3.1.0
 dynamodb =
     docker>=3.0.0
-    py-partiql-parser==0.4.0
+    py-partiql-parser==0.4.1
 dynamodbstreams =
     docker>=3.0.0
-    py-partiql-parser==0.4.0
+    py-partiql-parser==0.4.1
 ebs = sshpubkeys>=3.1.0
 ec2 = sshpubkeys>=3.1.0
 ec2instanceconnect =
@@ -205,15 +205,15 @@ resourcegroupstaggingapi =
     openapi-spec-validator>=0.2.8
     pyparsing>=3.0.7
     jsondiff>=1.1.2
-    py-partiql-parser==0.4.0
+    py-partiql-parser==0.4.1
 route53 =
 route53resolver = sshpubkeys>=3.1.0
 s3 =
     PyYAML>=5.1
-    py-partiql-parser==0.4.0
+    py-partiql-parser==0.4.1
 s3crc32c =
     PyYAML>=5.1
-    py-partiql-parser==0.4.0
+    py-partiql-parser==0.4.1
     crc32c
 s3control =
 sagemaker =

--- a/tests/test_dynamodb/test_dynamodb_statements.py
+++ b/tests/test_dynamodb/test_dynamodb_statements.py
@@ -198,3 +198,65 @@ class TestBatchExecuteStatement(TestCase):
             "TableName": "table1",
         } in items
         assert {"TableName": "table2", "Item": self.item1} in items
+
+
+@pytest.mark.aws_verified
+@dynamodb_aws_verified
+def test_execute_statement_with_all_clauses(table_name=None):
+    dynamodb_client = boto3.client("dynamodb", "us-east-1")
+
+    items = [
+        {
+            "pk": {"S": "0"},
+            "Name": {"S": "Lambda"},
+            "NameLower": {"S": "lambda"},
+            "Description": {"S": "Run code in under 15 minutes"},
+            "DescriptionLower": {"S": "run code in under 15 minutes"},
+            "Price": {"N": "2E-7"},
+            "Unit": {"S": "invocation"},
+            "Category": {"S": "free"},
+            "FreeTier": {"N": "1E+6"},
+        },
+        {
+            "pk": {"S": "1"},
+            "Name": {"S": "Auto Scaling"},
+            "NameLower": {"S": "auto scaling"},
+            "Description": {
+                "S": "Automatically scale the number of EC2 instances with demand",
+            },
+            "DescriptionLower": {
+                "S": "automatically scale the number of ec2 instances with demand"
+            },
+            "Price": {"N": "0"},
+            "Unit": {"S": "group"},
+            "Category": {"S": "free"},
+            "FreeTier": {"NULL": True},
+        },
+        {
+            "pk": {"S": "2"},
+            "Name": {"S": "EC2"},
+            "NameLower": {"S": "ec2"},
+            "Description": {"S": "Servers in the cloud"},
+            "DescriptionLower": {"S": "servers in the cloud"},
+            "Price": {"N": "7.2"},
+            "Unit": {"S": "instance"},
+            "Category": {"S": "trial"},
+        },
+        {
+            "pk": {"S": "3"},
+            "Name": {"S": "Config"},
+            "NameLower": {"S": "config"},
+            "Description": {"S": "Audit the configuration of AWS resources"},
+            "DescriptionLower": {"S": "audit the configuration of aws resources"},
+            "Price": {"N": "0.003"},
+            "Unit": {"S": "configuration item"},
+            "Category": {"S": "paid"},
+        },
+    ]
+
+    for item in items:
+        dynamodb_client.put_item(TableName=table_name, Item=item)
+
+    partiql_statement = f"SELECT pk FROM \"{table_name}\" WHERE (contains(\"NameLower\", 'code') OR contains(\"DescriptionLower\", 'code')) AND Category = 'free' AND Price >= 0 AND Price <= 1 AND FreeTier IS NOT MISSING AND attribute_type(\"FreeTier\", 'N')"
+    items = dynamodb_client.execute_statement(Statement=partiql_statement)["Items"]
+    assert items == [{"pk": {"S": "0"}}]

